### PR TITLE
Read-only course details

### LINKED ISSUE #133: Course details should be accessible read-only for student representatives in all stages
At the moment, for courses which are in one of the states "evaluated", "reviewed" or "published", student representatives cannot see the details (like lecturer, participants, and so on) of this course.

It is true that the details of a course MUST NOT be edited in those states. However, it is desirable to have read-only access to this information e.g. to check the list of lecturers and to identify potential permission issues arising from it ("X is proxy of Y, but why can't X access Y's course results? Is Y really the main lecturer [='Dozent' flag]?)

### PR DESCRIPTION
**Pull request by kateyy**
Published and archived courses can now be opened in a read-only form. For reviewed courses, a warning message will be displayed on the edit form. (Closes #133)

### CODE REVIEW

File: evap/staff/templates/staff_course_form.html [hunk 1]
```diff
@@ -64,10 +68,18 @@
                 {% if form.instance.state == "inEvaluation" %}
                     <div class="alert alert-warning">{% trans "You are editing a course, which is already in evaluation. Please note that only the students who did not evaluate yet will see your changes." %}</div>
                 {% endif %}
+                {% if form.instance.state == "evaluated" %}
+                    <div class="alert alert-warning">{% trans "You are editing a course, for which the evaluation already finished. Please be careful to not destroy any results." %}</div>
+                {% endif %}
+                {% if form.instance.state == "reviewed" %}
+                    <div class="alert alert-warning">{% trans "You are editing a course, which is already review. Please note that only the students who did not evaluate yet will see your changes." %}</div>
```
Comment (karyon) [inline evap/staff/templates/staff_course_form.html]: could you remove this and just do `if form.instance.state == "evaluated" or form.instance.state == "reviewed"` above?

File: evap/staff/templates/staff_course_form.html [hunk 2]
```diff
@@ -78,5 +90,7 @@
 
 {% block additional_javascript %}
     {% include "evap_course_edit_js.html" %}
-    <script type="text/javascript"> preventAccidentalClosing = true; </script>
+    <script type="text/javascript">
+        preventAccidentalClosing =  {{ form.instance.can_staff_edit|yesno:"true,false" }};
+    </script>
```
Comment (karyon) [inline evap/staff/templates/staff_course_form.html]: could you do 
```
{% if form.instance.can_staff_edit %}
    <script type="text/javascript"> preventAccidentalClosing = true; </script>
{% endif %}
```

File: evap/staff/templates/staff_single_result_form.html [hunk 3]
```diff
@@ -11,12 +11,16 @@
...
             </div>
         </div>
     </form>
 {% endblock %}
 
 {% block additional_javascript %}
-    <script type="text/javascript"> preventAccidentalClosing = true; </script>
+    <script type="text/javascript"> 
+        preventAccidentalClosing = {{ form.instance.can_staff_edit|yesno:"true,false" }};
+    </script>
```
Comment (karyon) [inline evap/staff/templates/staff_single_result_form.html]: same here

File: evap/staff/templates/staff_single_result_form.html [hunk 4]
```diff
@@ -11,12 +11,16 @@
 
         <div class="form-group well">
             <div class="col-sm-offset-2 col-sm-6">
-                <input type="submit" value="{% trans "Save course" %}" class="btn btn-primary" onclick="preventAccidentalClosing = false;"/>
+                {% if form.instance.can_staff_edit %}
+                    <input type="submit" value="{% trans "Save course" %}" class="btn btn-primary" onclick="preventAccidentalClosing = false;"/>
+                {% endif %}
```
Comment (karyon) [inline evap/staff/templates/staff_single_result_form.html]: maybe this if should be around the whole `<div class="form-group well">`?

File: evap/staff/views.py [hunk 5]
```diff
@@ -462,6 +456,10 @@ def helper_course_edit(request, semester, course):
     if form.is_valid() and formset.is_valid():
         if operation not in ('save', 'approve'):
             raise SuspiciousOperation("Invalid POST operation")
+
+        if (not course.can_staff_edit or course.is_archived) and operation == 'save':
+            raise SuspiciousOperation("Modifying this course is not allowed.")
```
Comment (karyon) [inline evap/staff/views.py]: i think this check should also apply for the `approve` operation, right? we don't want people to approve courses that they cannot edit.

File: evap/staff/views.py [hunk 6]
```diff
@@ -494,6 +492,9 @@ def helper_single_result_edit(request, semester, course):
     form = SingleResultForm(request.POST or None, instance=course, initial=initial)
 
     if form.is_valid():
+        if (not course.can_staff_edit or course.is_archived):
```
Comment (karyon) [inline evap/staff/views.py]: remove parentheses please

**Comment by karyon** [comment]
@kateyy do you know why i didn't review this two weeks ago? were there any major problems we talked about?

File: evap/staff/forms.py [hunk 7]
```diff
@@ -183,6 +192,11 @@ def __init__(self, *args, **kwargs):
 
         self.fields['questionnaires'].queryset = Questionnaire.objects.filter(is_for_contributors=True).filter(
             Q(obsolete=False) | Q(contributions__course=self.course)).distinct()
+        
+        if not self.course.can_staff_edit:
+            # form is used as read-only course view
+            for field in self.fields.values():
+                field.disabled = True
```
Comment (karyon) [inline evap/staff/forms.py]: before i forget it again, this should go into a method `def disable_all_fields(form):`

**Comment by karyon** [comment]
@kateyy any chance you'll complete this e.g. tomorrow or say, in the next two weeks? otherwise i'd take this over.

File: evap/staff/templates/staff_course_form.html (hunk 1 above)
Comment (kateyy) [inline evap/staff/templates/staff_course_form.html]: The warning messages are (intentionally) different, so I wouldn't change this.

Comment (janno42) [inline evap/staff/templates/staff_course_form.html]: `reviewed` comes after `evaluated`, the course is put into that state once all comments have been reviewed. the states are otherwise the same, so we should actually show the same message for both.

File: evap/staff/templates/staff_course_form.html (hunk 2 above)
Comment (kateyy) [inline evap/staff/templates/staff_course_form.html]: done

File: evap/staff/templates/staff_single_result_form.html (hunk 4 above)
Comment (kateyy) [inline evap/staff/templates/staff_single_result_form.html]: Yep. The same would be nice for staff_course_form.html, but I didn't find a solution for not checking all these ifs again around the div.

Comment (janno42) [inline evap/staff/templates/staff_single_result_form.html]: what do you mean by "all these ifs"?
what about this?:
```
{% if form.instance.can_staff_edit %}
    <div class="form-group well">
        <div class="col-sm-offset-2 col-sm-6">
            <input type="submit" value="{% trans "Save course" %}" class="btn btn-primary" onclick="preventAccidentalClosing = false;"/>
        </div>
    </div>
{% endif %}
```

Comment (janno42) [inline evap/staff/templates/staff_single_result_form.html]: oh, i see. nevermind, for staff_course_form we don't need it because the save button is always there.

File: evap/staff/views.py (hunk 5 above)
Comment (kateyy) [inline evap/staff/views.py]: Done

File: evap/staff/tests.py [hunk 8]
```diff
@@ -1089,7 +1088,6 @@ def test_raise_403(self):
         self.get_assert_403(semester_url + "import", "evap")
         self.get_assert_403(semester_url + "assign", "evap")
         self.get_assert_403(semester_url + "course/create", "evap")
-        self.get_assert_403(semester_url + "course/7/edit", "evap")
```
Comment (janno42) [inline evap/staff/tests.py]: why did you remove this?

File: evap/staff/templates/staff_course_form.html [hunk 9]
```diff
@@ -64,10 +68,18 @@
                 {% if form.instance.state == "inEvaluation" %}
```
Comment (janno42) [inline evap/staff/templates/staff_course_form.html]: please add an if around the well div. it probably will be quite long, but an empty div makes no sense.

File: evap/staff/views.py [hunk 10]
```diff
@@ -462,6 +456,10 @@ def helper_course_edit(request, semester, course):
     if form.is_valid() and formset.is_valid():
         if operation not in ('save', 'approve'):
             raise SuspiciousOperation("Invalid POST operation")
+
+        if (not course.can_staff_edit or course.is_archived) and operation in ('save', 'approve'):
```
Comment (janno42) [inline evap/staff/views.py]: the operation must be either `save` or `approve`at this point, so `if not course.can_staff_edit or course.is_archived:` would be enough.

File: evap/staff/templates/staff_course_form.html [hunk 11]
```diff
@@ -59,24 +61,33 @@
             </tbody>
         </table>
 
+        {% if state == "inEvaluation" or state == "evaluated" or state == "reviewed" or state == "new" or state == "prepared" or state == "editorApproved" or course.can_staff_edit %}
```
Comment (janno42) [inline evap/staff/templates/staff_course_form.html]: i'm sorry that i didn't completely think this through before. the states listed here are all states but `published` (`approved` is missing, but should actually be listed as well). so in all states in which the course can be edited we need this div. that means we can simply put a `{% if course.can_staff_edit %}` here.

The `Save` button can always be rendered (because we already checked for the `can_staff_edit`), the `Save and approve` button keeps the if with its three states.

**Comment by kateyy** [comment]
saving courses in evaluated or reviewed now triggers TransitionNotAllowed at reopen_evaluation...

**Comment by janno42** [comment]
`is_in_evaluation_period` is not a property and must be called `is_in_evaluation_period()`. this is currently wrong in two places in the code: `staff/views.py` line 491, and `evaluation/models.py` line 271.
when you fix that it should work.

**Comment by kateyy** [comment]
okay fixed.

**Comment by janno42** [comment]
when editing a published course, the disabled contribution formset does not render correctly (the selection fields with the contributors' names are too short). also, it is currently possible to add and delete forms in this formset.
please let me know if you don't want to look into that because it's too much trouble.

**Comment by kateyy** [comment]
``class="movable"`` only sets ``width: 10px``, so it's fixed now, but it's maybe not the best style to use this class for non-movable elements?

**Comment by janno42** [comment]
i'm okay with that class there. it's the element for moving stuff around even if it's not enabled ;)
do you want to look into the add and remove formset stuff?

**Comment by kateyy** [comment]
yep

**Comment by kateyy** [comment]
Is there a way to remove the "add another" without modifying the jquery.formset.js plugin?

**Comment by janno42** [comment]
you could simply not call the `make_form_sortable` in `evap_course_edit_js.html`, then you get rid of all the formset stuff (you might not even need the movable td then anymore).
the only thing left then is the delete checkbox which you need to hide.

**Comment by janno42** [comment]
it looks good now! thanks for fixing all the small thing i had to criticize.
could you now please squash at least some of the commits together (if you want they can all go into one)? then we're ready to merge.

**Comment by kateyy** [comment]
Done =)
